code/crates/core-driver/src/driver.rs

```diff
@@ -45,9 +45,23 @@ where
     /// The proposals to decide on.
     pub(crate) proposal_keeper: ProposalKeeper<Ctx>,
 
+    /// The evidence of equivocation (if any) after processing a proposal.
+    pub(crate) proposal_equivocation_evidence: Option<(
+        Ctx::Height,
+        Ctx::Address,
+        (SignedProposal<Ctx>, SignedProposal<Ctx>),
+    )>,
```

Comment:
Let's define a proper type for this instead of using a tuple

Reply:
By the way, can you clarify why we need these two new variables when we already have evidence maps in the proposal and vote keepers?